data_pull.py

```diff
@@ -53,33 +81,158 @@ def pull_data(self, search_criteria: str, delete_csv: bool = False):
-        download_dir = os.path.join(dirname, "data", "")
-        file_type = "*.csv"
-
         # Open Chrome web browser
         driver = webdriver.Chrome(service=service, options=options)
 
+        # Define directory where data will be downloaded and file type
+        dirname = os.path.abspath(os.path.dirname(__file__))
+        download_dir = os.path.join(dirname, 'data', '')
+
         # Enable Chromium driver to download files while in headless mode
-        params = {"behavior": "allow", "downloadPath": download_dir}
-        driver.execute_cdp_cmd("Page.setDownloadBehavior", params)
+        params = {'behavior': 'allow', 'downloadPath': download_dir}
+        driver.execute_cdp_cmd('Page.setDownloadBehavior', params)
 
         # Navigate to Redfin website
-        driver.get("http://www.redfin.com")
-        driver.implicitly_wait(1)
+        driver.get('http://www.redfin.com')
+        time.sleep(1.5)
+
+        return driver, download_dir
+
+    def listing_data(self, driver: webdriver, download_dir: str, search_criteria: str, home_types: Union[str, list],
+                     delete_csv: bool = False):
+        """
+        Attempts to download csv of Redfin real estate listing data for a specified search area
+        :param driver: Webdriver used to connect to Redfin website
+        :param download_dir: Directory where downloaded CSV files will go
+        :param search_criteria: Search criteria to be entered into Redfin search bar (city, state, address, etc.)
+        :param home_types: Type(s) of home of which to pull data for. Options include: house, townhouse, condo, land,
+        multi-family, mobile, co-op, and other. If selecting more than one home type, initialize param as a list
+        :param delete_csv: When set to True, deletes the CSV file that is downloaded as part of the data pulling process
+        :return: Pandas DataFrame containing real estate data from Redfin for the desired city & state
+        """
+        # Identify search bar on home page for entering a real estate search location and search for input city/state
+        # Finds first search bar on the home page, which is the one we're looking for
+        driver.find_element(By.CLASS_NAME, 'search-input-box').send_keys(f'{search_criteria}' + Keys.ENTER)
+        time.sleep(1.5)
+
+        # Click 'All Filters'; Set to show 'For Sale' listings only by default
+        driver.find_element(By.XPATH, '//*[@id="sidepane-header"]/div/div[1]/form/div[5]/div').click()
+        time.sleep(1.5)
+
+        # Select desired 'Home Type'
+        home_type_header = driver.find_element(By.XPATH, '//*[@id="filterContent"]/div/div[5]/div[1]/div/span')
+        driver.execute_script("return arguments[0].scrollIntoView();", home_type_header)
+        time.sleep(1.5)
+
+        if type(home_types) == str:     # If user selects single home type (string), convert it to a list
+            home_types = [home_types]
```

Comment:
That's how I had it originally, and would be fine if I'm truly the only one who ever uses it. I made it a bit more robust that way in case anyone else wanted to clone/fork the repo and use it themselves. I'll have to sit on this one.

Reply:
In that case you have two alternative options:

### Method Overloading

```
def listing_data(self, driver: webdriver, download_dir: str, search_criteria: str, home_types: list, ...):
      {existing code}

def listing_data(self, driver: webdriver, download_dir: str, search_criteria: str, home_types: str, ...):
       listing_data(self, driver, download_dir, search_criteria, [home_types],....)
```

What I've done is make it so both string and list are acceptable data types, but when string is used, I convert it to a singleton list (list with one element) while making the call with the List only as the argument. 

### Varargs
Varargs allows you to pass any number of objects/data structures as arguments. 
In Java you'd declare a method like this:
```
public void listingData(String... houseTypes) {
}
```
The variable `houseTypes` if using varargs must be the last parameter in the method signature and within the method is represented as an Array. This means I could call it in any of these ways and it would be valid:
```
listingData();
listingData("single-family");
listingData("single-family", "duplex"); 
```

In Python I'm not certain it has to be the last parameter in the method signature, but you just have to say `*house_types: str`. You could swap in the HouseTypes Enum signature instead of str here as well if you go that route. See the *args section of this article for an example: https://www.geeksforgeeks.org/args-kwargs-python/